### ISSUE TITLE
loklak Tweets - compose tweets with new rich content attachment categories, e.g. embedded maps or code with loklak

### ISSUE THREAD

**Issue by Orbiter** [loklak/loklak_webclient#63]
This is an umbrella feature for the project definition in https://github.com/loklak/loklak_webclient/issues/61
The idea is, that users like to tweet rich data but don't have support for such attachments from twitter. Examples are:
- identify and attach exact geographic coordinates and maps
- attach video and/or audio content
- attach larger texts with simple markdown (i.e. https://guides.github.com/features/mastering-markdown/) as used in github or maybe wikitext, code citations (with code pretty-print) and other typical text rendering
- attach better images (i.e. animated gif or images with EXIF data attached)

Tweets with such attachments shall be stored within loklak with full rich data but will also be cross-posted to twitter with their metadata attached in an 'encapsulated' form (i.e. geographic data as OSM maps with engraved coordinates or large markup-texts as attached images etc.) Details will be explained in sub-features.

This feature has an important requirement: authentication of the user and using loklak as twitter app. Because this is not available right now, we will start to implement these features without cross-posting and add that later.

Details will be defined in sub-features

**Comment by mariobehling** [comment]
@sudheesh001 Create a list of features and add information about technology and milestones. Please include the following:

a) Feature List (add it to Feature.md)
b) Project Plan with possible milestones (Which features will be implemented at what time)
c) Technologies used

Proposed Milestones

    Milestone 1, Due by June 10, 2015
    Milestone 2, Due by June 26, 2015
    Milestone 3, Due by July 25, 2015
    Milestone 4, Due by August 21, 2015

**Comment by sudheesh001** [comment]
### Feature List
1. Enable the application and user capability to post tweets to twitter
2. Enable geo location and tweets coupled with geolocation to be posted to twitter
3. Enable displaying the geo location marker on our application via open street maps.
4. Integrate markdown support to tweets and send them to loklak_server, retrieve the image and post to twitter
5. Attaching video / audio content of the application
6. Have a feature of asking loklak for statistics via twitter itself and make loklak app reply to that
7. Daily tweet about a specific trending search query on twitter and required info to twitter.
8. Widgets that could be applied on wall and other search pages as required.

### Project plan and milestones

Milestone 1 (Twitter basics integrations)
- Integrate posting tweets to twitter (#92)
- Integrate allow geo location of the user to be found and tweet along with that information (#93)

Milestone 2 (Integrations support)
- Markdown support (#94)
- Image &GIF uploads (#95)
- Video and Audio playback support directly. Will need `loklak_server` support
- Video and Audio uploads sent to loklak_server to return a link. (#97)

Milestone 3 (Data mining and intelligence reporting / infographics => infotweets)
- Similar to how inforgraphics convey a lot of content, loklak_server hosts up a lot of information for each query users are making, it'd be a great addon to send out the results of the tweets searched for and the ones with highest statistics directly as a post to twitter at the EOD. Similar to SumAll app on Twitter
- Allow text files and other uploaded files to be opened and read on server, screened back to the frontend like an image so that the image can be posted to twitter.

Milestone 4 (Widgets and other integrations)
- Implement a few widgets that organizers / walls need, like a widget to show current speaker, next speaker, lunch time, dinner time, break etc.., or breakout rooms / different events in parallel rooms. I think this feature widget is required.
- Tweaks and testing of the system.

### Technologies used

Nodejs server
Passport JS
Twitter node npm modules
Mining via python worker roles (Still yet to decide how to go about this and feature number 7) (Coming soon)

**Comment by Orbiter** [comment]
To provide a storage back-end for rich tweets the loklak_server now provides an 'asset' servlet.
Documentation is here: https://github.com/loklak/loklak_server/issues/42#issuecomment-111807139

**Comment by sudheesh001** [comment]
Personal tasks to do
- [ ] Fire up the blogs and link the blog to the personal website.
- [ ] Reach out to more developers and audience.
- [ ] Build a project presentation so that I can pitch for it in upcoming tech meetups nearby.

**Comment by Orbiter** [comment]
Here is an overview of the process to post rich tweets:
- let the user compose the message
- convert non-tweetable content into an image
- tweet the message with the rich-content as image
- IMPORTANT: save the tweet id of that tweet posted to twitter
- compose a tweet for loklak_server
- set the field `canonical_id` according to https://github.com/loklak/loklak_server/issues/64 with the content of the recently posted message to twitter
- submit the tweet with rich content to loklak_server using the /api/push.json api
- submit the assets (rich attachments of the message) to /api/asset

**Comment by sudheesh001** [comment]
The plan for the upcoming deadlines are:
### Milestone 3
1. Implementing the deletion functionality from the loklak interface. #306 
2. Page showing followers and following information at a user profile `/<user>/<followers>` and `/user/<following>` hoping that the /user is built by the account creation method view.
3. Cleaning up the unused libraries in npm which are making the repositories bulky and trigger errors during set up while doing it on new machines.
4. Allow code as markdown to be posted to twitter directly via the loklak app. (Posting rich tweets)

### Milestone 4
1. Extended live tweet validations
2. Cleaning up the code (ambitious to fix it by Milestone 3 so that the milestone 4 doesn't need to have this)
3. Post on behalf of organization members which would allow Organizational accounts to be built via loklak so that the organization owner can allow verified contributors to post via this account
4. Integrate with SMTP and send emails to the user email about notifications for such organizational accounts.
5. User profiles view for `loklak.net/<username>`

**Comment by mariobehling** [comment]
Please provide a list for feature issues for the milestones until 4.5
 Milestone 3, Due by July 25, 2015
 Milestone 3.5 Sprint, Due by August 12, 2015 
 Milestone 4, Due by August 21, 2015
 Milestone 4.5, Due by August 31, 2015

**Comment by mariobehling** [comment]
@sudheesh001 Currently your list of features in the milestones looks like a list of random ideas. This is not helping us to achieve the goal to make loklak into a service that can be used by the end of the summer. Let's go step by step. We would like to see a list of features from you first that are related to "tweets". Please fill in the Google doc I sent to you.

**Comment by mariobehling** [comment]
@sudheesh001 I have updated your plan and took out things out of scope. I also added goals for each milestones. Let's focus first on the current milestone. Do you think you can implement all the issues in the current milestone 3?

Other than that: I took out "Page showing followers and following information at a user profile" as this is currently assigned to Aneesh while he is taking care of accounts related work.

**Comment by Orbiter** [comment]
please retrieve the follower list from loklak_server, documentation is online at http://loklak.org/api.html#account .. please scroll down to "Retrieval of Twitter User Account Data" for details
